Foundation of New Product Collection Block

imanish003 opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This PR introduces the foundations for a new Product Collection block. It includes the creation of two new blocks:
1. Product Collection
2. Product Template

The Product Collection block will replace the existing Products (Beta) block. When the Product Collection block is added to the Editor, it will display all products in a grid view. On the frontend, all products should be displayed, and pagination should function as expected.

<!-- Reference any related issues or PRs here -->

Fixes #9267


### Testing

1. Create a new post 
2. Add the Product Collection block to the post.
3. Verify that the UI of the Product Collection block matches that of the existing Products Beta block in both the editor and the frontend.
4. Test the pagination functionality on the frontend to ensure it works as expected.

Here is a quick demo:

https://user-images.githubusercontent.com/16707866/236784525-9bb93c6d-e538-40d3-9524-4d71f91d4431.mov

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

### Changelog

> Foundation of New `Product Collection` Block

imanish003 committed:
Add Products Collection block scaffolding

This commit adds the initial scaffolding for the new Products Collection block. It includes the creation of new files (block.json, edit.tsx, index.tsx and ProductsCollection.php) and additions to existing files (webpack-entries.js and BlockTypesController.php). The block is marked as experimental and currently displays a static "Product collection" text in both the editor and the frontend.

imanish003 committed:
Add Product Template block and integrate with Products Collection block

This commit introduces the new Product Template block and integrates it with the existing Products Collection block. The changes include:
- Creation of the Product Template block, including its block.json, edit.tsx, icon.tsx, index.tsx, and save.tsx files.
- Modification of the Products Collection block, adding an icon.tsx file and updating its edit.tsx and index.tsx files.
- Updates to the webpack-entries.js file to include the new Product Template block.
- Addition of the ProductTemplate class in the src/BlockTypes directory.
- Inclusion of the ProductTemplate class in the src/BlockTypesController.php file.

imanish003 committed:
Enhance product-template block with context and styling

This commit adds various enhancements to the product-template block. It includes:

- Introduce `usesContext` and `supports` properties to the block.json file
- Update the edit.tsx file to use BlockContextProvider and add query logic
- Add an editor.scss file for styling the product-template block in the editor
- Extend the products-collection block.json file with new properties
- Modify the products-collection edit.tsx file to include instanceId and useEffect

These changes improve the product-template block by providing better context handling and styling options.

imanish003 committed:
Add 'woocommerce/product-template' to supported inner blocks and improve product-template editing

This commit adds 'woocommerce/product-template' as a supported inner block for various product elements, including button, image, rating, sale-badge, SKU, and stock-indicator. It also improves the editing experience for the product-template block by memoizing the block previews and rendering them more efficiently. This should lead to a smoother editing experience and better performance in the block editor.

imanish003 committed:
Add product title and summary variations for Products Collection block

- Add default spacing between product elements in `style.scss`
- Register product title and summary element variations in `products-collection` block
- Create utility for registering element variations
- Extend core elements with WooCommerce namespace

imanish003 committed:
Set inherit to false by default and other improvements

This commit introduces several changes to the ProductsCollection block:
1. Set the "inherit" property to false in block.json, disabling inheritance by default.
2. Remove the default styles for inner blocks in the edit.tsx file.
3. Add a save function in the index.tsx file, and create a save.tsx file to handle saving the block.
4. Update the ProductTemplate.php file to properly render the block content based on the changes.
5. Remove the render function from the ProductsCollection.php file, as it is no longer necessary.

These changes improve the functionality and flexibility of the ProductsCollection block, allowing for better customization and control over the block's appearance and behavior.

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/9267-foundation-of-products-collection

imanish003 committed:
Fix pagination issue

imanish003 committed:
Minor code quality improvments

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/9267-foundation-of-products-collection

imanish003 committed:
Register product blocks only in experimental builds

- Wrap registerBlockType calls in product-template and products-collection with isExperimentalBuild check to enable block registration only in experimental builds.
- Update the default value for the perPage property in products-collection/block.json from null to 9, setting a default display of 9 items per page.

imanish003 committed:
Merge branch 'trunk' into add/9267-foundation-of-products-collection

imanish003 committed:
Add experimental flag to PHP file

imanish003 committed:
Update documentation for feature flags

- Add Products Collection and Product Template blocks to the list of experimental flags in the feature-flags-and-experimental-interfaces.md file.
- Include references to PHP and webpack flags for both blocks.

roykho commented:
Some of the frontend JS files are not loaded correctly:
![OcD7m9.png](https://github.com/woocommerce/woocommerce-blocks/assets/2132595/4fb28fac-7832-42f8-bc2c-0ee202e41373)

When I put the `Products (beta)` block and `Products Collection` block side by side, I can see different loading orders of the products. I am not sure if this is an issue or not. Looking at your example, they're the same. But on mine they are not.
![hbgfyc.png](https://github.com/woocommerce/woocommerce-blocks/assets/2132595/f1fbf790-2d3c-4752-b78a-a1ab86c4d21e)

imanish003 committed:
Change default order and orderBy values in block.json

- Update the default 'order' value from 'desc' to 'asc'.
- Update the default 'orderBy' value from 'date' to 'title'.

imanish003 commented:
Hi @roykho, thanks for the feedback 🙇 

> Some of the frontend JS files are not loaded correctly:

I couldn't replicate this issue. Could you please try running `rm -rf build` to remove the build directory, and then execute `npm start`? This should resolve the error. 🙂

> When I put the Products (beta) block and Products Collection block side by side, I can see different loading orders of the products. I am not sure if this is an issue or not. Looking at your example, they're the same. But on mine they are not.

Apologies for the confusion. In fact, I manually changed the order for the Products beta block for comparison purposes. I selected `Newest` in the dropdown. Currently, the order is hardcoded in the block.json file [here](https://github.com/woocommerce/woocommerce-blocks/blob/94a75c67b31c1afbea4dc7ec3c597dd5cd60c534/assets/js/blocks/products-collection/block.json#L22-L23). However, we plan to add a dropdown in the sidebar settings, allowing users to modify the order.

<img width="271" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/45123b23-4d3a-4e7e-8348-f2112837da6b">

I've just pushed [aace308](https://github.com/woocommerce/woocommerce-blocks/pull/9352/commits/aace3080edfc85774dd75d53a4bd524fa618a310) to align the default ordering with the Products beta block. Now, both blocks will display the same order by default. 🙂

kmanijak commented on assets/js/blocks/product-template/index.tsx:
```diff
@@ -0,0 +1,24 @@
+
+if ( isExperimentalBuild() ) {
+	registerBlockType( metadata.name, blockConfig );
+}
```
You can use `registerExperimentalBlockType` function available in:

`import { registerExperimentalBlockType } from '@woocommerce/block-settings';`

Although it's basically doing the same thing, so leaving the decision up to you.

imanish003 replied:
I'm going to go ahead and use registerExperimentalBlockType to make the code a bit cleaner. If there's an abstraction available, I think it's better to use it. Thanks Karol.

kmanijak reviewed (commented):
`assets/js/blocks/product-template/types.ts` is an empty file, so I suggest removing it for now.

roykho commented:
Even after removing the build directory and re-running `npm start` I am still seeing the same missing `JS` files error in console.

kmanijak commented:
> Even after removing the build directory and re-running npm start I am still seeing the same missing JS files error in console.

@imanish003, I think this piece of code is required to prevent the `*-frontend.js` file to be fetched for:

```
	/**
	 * Get the frontend script handle for this block type.
	 *
	 * @param string $key Data to get, or default to everything.
	 */
	protected function get_block_type_script( $key = null ) {
		return null;
	}
```

I remember adding it to the block's class for the same reason. Here are examples:
- [ProductResultsCount.php](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/BlockTypes/ProductResultsCount.php#L24)
- [CatalogSorting.php](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/BlockTypes/CatalogSorting.php#L54)

imanish003 commented:
@kmanijak Oh yes, thank you 🙏🏻! I believe implementing your suggested solution will resolve the issue.

imanish003 committed:
Refactor experimental block registration and remove unused file

- Replace `isExperimentalBuild()` with `registerExperimentalBlockType()` in product-template/index.tsx.
- Remove unused file types.ts in product-template directory.
- Add `get_block_type_script()` function to ProductTemplate.php and ProductsCollection.php to return null.

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/9267-foundation-of-products-collection

imanish003 committed:
Update variation names in product summary and product title elements

Updated variation names for product summary and product title elements to match the new products-collection namespace. The previous variation names used the product-query namespace which will be deprecated. This change ensures that the correct variation names are used for these elements in the products-collection block.

imanish003 commented:
@roykho 
> Even after removing the build directory and re-running npm start, I am still encountering the same missing JS files error in the console.

Oh, my bad! I was checking on the Editor side, but I am able to reproduce it on the frontend side. I was able to fix it by following Karol's suggestion in the comment above.

@kmanijak, I have also implemented the other minor improvements you suggested. Could you please take another quick look? 🙏🏻

roykho commented on assets/js/blocks/product-template/index.tsx:
```diff
@@ -0,0 +1,22 @@
+/**
+ * External dependencies
+ */
+import { BlockConfiguration, registerBlockType } from '@wordpress/blocks';
```
Perhaps remove this `registerBlockType` since it is not used.

roykho reviewed (approved):
Just a small comment but not a blocker. Otherwise LGTM. The frontend JS errors are now gone. Good job!

kmanijak reviewed (changes requested):
I just spotted one tiny, but significant thing 😅 

Block is supposed to be named Product Collection, not Products Collection (singular, not plural).

From ref: pdnLyh-3uF-p2
> Rename from Products to Product Collection

Let's unify it now, since it's a kick-off PR, so that would mean:
- changing the block name and its references
- changing the directory in which the block is stored.

imanish003 commented:
Hi @kmanijak, I actually had the intention of confirming this with @shaunandrews, but I forgot to do so. 🤦‍♂️

@shaunandrews, just to confirm, we want to name the new block as `Product Collection` instead of `Products Collection`, correct? 🙂

imanish003 committed:
Rename 'Products Collection' to 'Product Collection'

- Renamed all instances of "products-collection" to "product-collection" across multiple files.
- Adjusted related types and method calls to match the new naming convention.
- Updated documentation and feature flags to reflect the name change.
- Made necessary changes in `BlockTypesController.php` and `bin/webpack-entries.js`.

imanish003 commented:
Hey @kmanijak, thank you for pointing this out. I have made the necessary changes in [f3453cc](https://github.com/woocommerce/woocommerce-blocks/pull/9352/commits/f3453ccdc15014b85c93988bb426edb906506010). Could you please take a quick look?

kmanijak reviewed (approved):
Thanks for all the changes! 💪 Please also change Products Collection to Product Collection in the PR title and changelog. 🙏 

One last thing I found is that Product Template should not be available in the inserter. But anyway, I'm pre-approving!

imanish003 committed:
Hide product-template block from inserter

- Added "woocommerce/product-collection" to the list of parent blocks in `product-template/block.json`.
- Added `"inserter": false` to the "supports" section to disable the option of inserting this block through the editor inserter component.

imanish003 committed:
Merge branch 'trunk' into add/9267-foundation-of-products-collection

imanish003 commented:
Hey @kmanijak, good catch! 🙌🏻

I have made the changes to the title and description of this PR. Additionally, after the changes made in commit [9bc39b2](https://github.com/woocommerce/woocommerce-blocks/pull/9352/commits/9bc39b22482f4f6f9025d9d30133140c971422de), the Product Template block will no longer be available in the inserter. 🤝

imanish003 committed:
Fix Eslint error

imanish003 committed:
Merge branch 'trunk' into add/9267-foundation-of-products-collection
